src/widgets/CategoryWidget.js

```diff
@@ -24,48 +25,73 @@ import { AggregationTypes } from './AggregationTypes';
   const { column } = props;
   const [categoryData, setCategoryData] = useState(null);
   const [selectedCategories, setSelectedCategories] = useState([]);
-  const [loading, setLoading] = useState(true);
+  const [loading, setLoading] = useState(false);
   const dispatch = useDispatch();
   const viewport = useSelector((state) => props.viewportFilter && state.carto.viewport);
   const source = useSelector((state) => selectSourceById(state, props.dataSource) || {});
-  const { data, credentials } = source;
+  const vF = useSelector((state) => state.carto.viewportFeatures);
+  const { data, credentials, sourceType: dataExtractMode } = source;
 
   useEffect(() => {
-    const abortController = new AbortController();
-    if (
-      data &&
-      credentials &&
-      (!props.viewportFilter || (props.viewportFilter && viewport))
-    ) {
-      const filters = getApplicableFilters(source.filters, props.id);
-      setLoading(true);
-      getCategories({
-        ...props,
-        data,
-        filters,
-        credentials,
-        viewport,
-        opts: { abortController },
-      })
-        .then((data) => {
-          setCategoryData(data);
-          setLoading(false);
-        })
-        .catch((error) => {
-          if (error.name === 'AbortError') return;
-          if (props.onError) props.onError(error);
-        });
-    } else {
-      setCategoryData(null);
+    if (dataExtractMode === 'TileLayer' && props.viewportFilter) {
+      throw new Error(`"viewportFilter" should be false if Source Type is "${AggregationTypes.TILE_LAYER}"`);
+    }
+  }, [props.viewportFilter]);
+
+  useEffect(() => {
+    const {dataSource, operation, operationColumn, column} = props;
+
+    if (dataExtractMode === 'TileLayer') {
+      const targetFeatures = vF[dataSource];
```

Comment:
Instead of doing this:
```javascript
const {layer, dataSource} = props;
if (layer) {
     getCategoriesFromLayer(layer, ...)
} else {
   // old code using SQL api
   getCategoriesFromSQL(...)
}
```
getCategoriesFromLayer:
```javascript
function getCategoriesFromLayer(layer, ...) {
 const features = layer. getRenderedFeatures();
 const groupedFeatures = ...
 return groupedFeatures